Ensure coupon codes are uppercase

alecritson opened the pull request:
Should Close #2252 by ensuring coupons are stored and retrieved uppercase regardless of where they are introduced.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Coupon codes are now automatically converted to uppercase when applied to carts and discounts.

* **Bug Fixes**
  * Adjusted tests to expect coupon codes in uppercase format.

* **Chores**
  * Removed unnecessary test group annotation.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

alecritson committed:
Ensure coupon code is uppercase

actions-user committed:
chore: fix code style

alecritson committed:
Handle null

alecritson committed:
fix test

alecritson commented:
> @alecritson I think this is OK, as long as we restrict coupon codes to ASCII-only. I'm guessing we don't have that in place currently.

We don't have that in place currently afaik so it would be it's own feature if required.

glennjacobs commented:
> > @alecritson I think this is OK, as long as we restrict coupon codes to ASCII-only. I'm guessing we don't have that in place currently.
> 
> We don't have that in place currently afaik so it would be it's own feature if required.

It's not really a separate feature. This PR can't be merged if we're uppercasing non-ASCII characters.

alecritson committed:
Move to casting class

alecritson committed:
Merge branch '1.x' into ensure-coupon-codes-are-uppercase

actions-user committed:
chore: fix code style

alecritson committed:
Fix tests

alecritson committed:
Merge branch 'ensure-coupon-codes-are-uppercase' of github.com:lunarphp/lunar into ensure-coupon-codes-are-uppercase

alecritson committed:
test fixes

actions-user committed:
chore: fix code style

alecritson committed:
Update to test set method

alecritson committed:
Merge branch 'ensure-coupon-codes-are-uppercase' of github.com:lunarphp/lunar into ensure-coupon-codes-are-uppercase

alecritson committed:
Fix tests

actions-user committed:
chore: fix code style

glennjacobs commented:
@alecritson can you explain the recent updates for me?

glennjacobs commented:
> > I think this is OK, as long as we restrict coupon codes to ASCII-only. I'm guessing we don't have that in place currently.
> 
> > It's not really a separate feature. This PR can't be merged if we're uppercasing non-ASCII characters.
> 
> @glennjacobs So the changes should make sure this is handled, as the tests should demonstrate.

I think I've got this wrong. I'll do some research and come back.

alecritson committed:
Simplify and rename

alecritson committed:
Merge branch 'ensure-coupon-codes-are-uppercase' of github.com:lunarphp/lunar into ensure-coupon-codes-are-uppercase

alecritson commented:
> @alecritson
> 
> So initially I was concerned that uppercasing everything would fundamentally change the code. E.g. `ß` would become `SS`.
> 
> However, using `strtoupper()`only uppercases ASCII characters, so I think we were OK in the first place, doh!
> 
> E.g.
> 
> ```
> Str::upper('🏆promo');     // '🏆PROMO'
> ```
> 
> Sorry, I think we simply need to revert the recent changes...

Reverted but kept the casting class, just renamed for future scaling if needed.
